### PR TITLE
hosting:core module config

### PR DESCRIPTION
**Pull request by Rick-Anderson**

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@shirhatti  config needs to live it it's own doc. Please review.

**Comment by shirhatti** [comment]
@Rick-Anderson I just made a couple of changes. Please review

**Comment by Rick-Anderson** [comment]
@shirhatti  looks great. Can you give me the squirrel?

**Comment by shirhatti** [comment]
I thought you are going to add in the `app_offline` stuff as well?

**Comment by Rick-Anderson** [comment]
@shirhatti sorry, added app_offline (https://github.com/aspnet/Docs/issues/1189 wasn't linked to this).

**Comment by Rick-Anderson** [comment]
@Shirhatti add note you must create log file and app_offline. Anything else?

**Comment by Rick-Anderson** [comment]
@Shirhatti added note you must create log file and app_offline. Anything else?

File: aspnet/hosting/aspnet-core-module.rst [hunk 1]
```diff
@@ -0,0 +1,155 @@
+ASP.NET Core Module Configuration Reference
+=============================================
+
+By `Rick Anderson`_ and `Sourabh Shirhatti`_
+
+In ASP.NET Core, the web application is hosted by an external process outside of IIS. The ASP.NET Core Module is an IIS 7.5+ module which is responsible for process management of  ASP.NET Core http listeners and to proxy requests to processes that it manages. This document provides an overview of how to configure the ASP.NET Core Module for shared hosting of ASP.NET Core.
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Nit: extra space between "of" and "ASP.NET"

File: aspnet/hosting/aspnet-core-module.rst [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+|                           | |                                                  |
+|                           | | Specifies the relative or absolute file path for |
+|                           | | which **stdout** and **stderr** from the process |
+|                           | | specified in **processPath** will be logged.     |
+|                           | | Relative paths are relative to the root of the   |
+|                           | | site. Any path starting with '.' will be         |
+|                           | | relative to the site root and all other paths    |
+|                           | | will be treated as absolute paths.               |
+|                           | |                                                  |
+|                           | | The default value is ``httpplatform-stdout``.    |
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: This should be `aspnetcore-stdout` now.

File: aspnet/hosting/aspnet-core-module.rst [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+|                           | | relative to the site root.                       |
+|                           | |                                                  |
+|                           | | There is no default value.                       |
++---------------------------+----------------------------------------------------+
+| arguments                 | | Optional string attribute.                       |
+|                           | |                                                  |
+|                           | | Arguments to the executable or script            |
+|                           | | specified in  **processPath**   .                |
+|                           | |                                                  |
+|                           | | There is no default value.                       |
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Nit: The schema specifies an empty string default value, so I guess that's the same thing.

File: aspnet/hosting/aspnet-core-module.rst [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+| startupTimeLimit          | | Optional integer attribute.                      |
+|                           | |                                                  |
+|                           | | Duration in seconds for which the                |
+|                           | | the handler will wait for the executable or      |
+|                           | | script to start a process listening on           |
+|                           | | the port. If this time limit is exceeded,        |
+|                           | | the handler will kill the process and try to     |
+|                           | | launch it again **startupRetryCount** times.     |
+|                           | |                                                  |
+|                           | | The default value is 10.                         |
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Default value is 120 according to the file I have here.

File: aspnet/hosting/aspnet-core-module.rst [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+|                           | |                                                  |
+|                           | | There is no default value.                       |
++---------------------------+----------------------------------------------------+
+| startupTimeLimit          | | Optional integer attribute.                      |
+|                           | |                                                  |
+|                           | | Duration in seconds for which the                |
+|                           | | the handler will wait for the executable or      |
+|                           | | script to start a process listening on           |
+|                           | | the port. If this time limit is exceeded,        |
+|                           | | the handler will kill the process and try to     |
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Nit: "attempt" perhaps better than "try"

File: aspnet/hosting/aspnet-core-module.rst [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+|                           | |                                                  |
+|                           | | Specifies the number of times the process        |
+|                           | | specified in **processPath** is allowed to crash |
+|                           | | per minute. If this limit is exceeded,           |
+|                           | | the handler will stop launching the              |
+|                           | | process for the remainder of the minute.         |
+|                           | |                                                  |
+|                           | | The default value is 10.                         |
++---------------------------+----------------------------------------------------+
+| requestTimeout            | | Optional timespan  attribute.                    |
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Nit: Extra space

File: aspnet/hosting/aspnet-core-module.rst [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+|                           | |                                                  |
+|                           | | Specifies the duration for which the             |
+|                           | | ASP.NET Core Module will wait for a response     |
+|                           | | from the process listening on                    |
+|                           | | %HTTP_PLATFORM_PORT%.                            |
+|                           | |                                                  |
+|                           | | The default value is "00:02:00".                 |
++---------------------------+----------------------------------------------------+
+| stdoutLogEnabled          | | Optional Boolean  attribute.                     |
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Nit: Extra space

File: aspnet/hosting/aspnet-core-module.rst [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+|                           | |                                                  |
+|                           | | If  true, the token will be forwarded to the     |
+|                           | | child process listening on %HTTP_PLATFORM_PORT%  |
+|                           | | as a header 'X-IIS-WindowsAuthToken' per request.|
+|                           | | It is the responsibility of that process to call |
+|                           | | CloseHandle on this token per request.           |
+|                           | |                                                  |
+|                           | | The default value is false.                      |
++---------------------------+----------------------------------------------------+
+
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: The section above didn't include `shutdownTimeLimit` (integerrange, 0-600, default: 10) and `processesPerApplication` (integer range, 1-100, default: 1).

File: aspnet/hosting/aspnet-core-module.rst [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+Child Elements
+^^^^^^^^^^^^^^^
+
++---------------------------+----------------------------------------------------+
+| Attribute                 |     Description                                    |
++===========================+====================================================+
+| environmentVariables      | | Configures **environmentVariables** collection   |
+|                           | | for the process specified in **processPath**.    |
++---------------------------+----------------------------------------------------+
+
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: The section didn't include `recycleOnFileChange`
```xml
<element name="recycleOnFileChange">
  <collection addElement="file" clearElement="clear">
    <attribute name="path" type="string" required="true" validationType="nonEmptyString" expanded="true"/>
  </collection>
</element>
```

File: aspnet/hosting/aspnet-core-module.rst [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+Once the *app_offline.htm* file is removed, the next request will load the application and the application will respond to requests.
+
+
+ASP.NET Core Module configuration examples
+------------------------------------------
+
+.. _log-redirection:
+
+Log creation and redirection
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Capitalization?

File: aspnet/hosting/aspnet-core-module.rst [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For logs to be saved, you must create the log directory. The ASP.NET Core Module can redirect ``stdout`` and ``stderr`` logs to disk by setting the ``stdoutLogEnabled``  and ``stdoutLogFile`` properties of the ``aspNetCore`` attribute. Logs are not rotated. It is the responsibilty of the hoster to limit the disk space the logs consume.
+
+.. literalinclude:: aspnet-core-module/sample/web.config
+  :language: xml
+  :lines: 12-16,20
+
+
+Setting environment variables
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Capitalization?

File: aspnet/hosting/aspnet-core-module.rst [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+| Attribute                 |     Description                                    |
++===========================+====================================================+
+| environmentVariables      | | Configures **environmentVariables** collection   |
+|                           | | for the process specified in **processPath**.    |
++---------------------------+----------------------------------------------------+
+
+ASP.NET Core Module app offline
+-------------------------------
+
+If you place a file with the name  *app_offline.htm* at the root of a web application directory, the ASP.NET Core Module will shut-down the application and stop processing any new incoming requests for that application. The ASP.NET Core Module will respond to all requests for dynamic pages in the application by sending back the content of the  *app_offline.htm*  file. For example, you might want to return a "site under construction" or a "down for maintenance" page.
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Nit: Are "any" and "new" unnecessary? ... i.e., an incoming request will be declined whether new or not ... and the last mention of "application" might be superfluous, as well.

> ... the ASP.NET Core Module will shut-down the application and stop processing incoming requests.

Issue #2 ...

> For example, you might want to return a "site under construction" or a "down for maintenance" page.

This doesn't seem like it's an example of the behavior just described. It's more of a tip, right? Should it say something like ...

> To avoid returning the default content of *app_offline.htm* , you might want to return a "site under construction" or a "down for maintenance" page.

It might also be a good idea to say how to do that. Tell the dev to modify the content of the file itself.

Also note that this language is dangerously similar to it's source: http://weblogs.asp.net/scottgu/426755 ... Personally, I'd rephrase a bit more to avoid needing a citation (but I also realize that Scott's blog is probably owned my MS, in which case it would be a courtesy to cite him).

File: aspnet/hosting/aspnet-core-module.rst [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+Once the *app_offline.htm* file is removed, the next request will load the application and the application will respond to requests.
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: These clauses are a little funky at the end, since we sort of expect not to find an independent clause back there. How about ...

> Once the *app_offline.htm* file is removed, the next request will load the application, which will then respond to requests.

File: aspnet/hosting/aspnet-core-module.rst [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+^^^^^^^^^^^^^^^
+
++---------------------------+----------------------------------------------------+
+| Attribute                 |     Description                                    |
++===========================+====================================================+
+| environmentVariables      | | Configures **environmentVariables** collection   |
+|                           | | for the process specified in **processPath**.    |
++---------------------------+----------------------------------------------------+
+
+ASP.NET Core Module app offline
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: Capitalization? or list the file explicitly ...

> Using *app_offline.htm* with the ASP.NET Core Module

File: aspnet/hosting/aspnet-core-module.rst [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+ASP.NET Core Module configuration examples
+------------------------------------------
+
+.. _log-redirection:
+
+Log creation and redirection
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For logs to be saved, you must create the log directory. The ASP.NET Core Module can redirect ``stdout`` and ``stderr`` logs to disk by setting the ``stdoutLogEnabled``  and ``stdoutLogFile`` properties of the ``aspNetCore`` attribute. Logs are not rotated. It is the responsibilty of the hoster to limit the disk space the logs consume.
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: `<aspNetCore>` ... is an "element" or "section" (in schema lingo) and the `stdoutLogEnabled` and `stdoutLogFile` are attributes.

> by setting the ``stdoutLogEnabled``  and ``stdoutLogFile`` attributes of the ``aspNetCore`` element.

File: aspnet/hosting/aspnet-core-module.rst [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+.. literalinclude:: aspnet-core-module/sample/web.config
+  :language: xml
+  :lines: 12-16,20
+
+
+Setting environment variables
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ASP.NET Core Module allows you specify environment variables for the process specified in the ``processPath`` setting by specifying them in ``environmentVariables`` child attribute to the ``aspNetCore`` attribute. The example below illustrates how you would use it.
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: I think these are both elements.

> specifying them in ``environmentVariables`` child element to the ``aspNetCore`` element.

... and this might not be necessary at all ...

> The example below illustrates how you would use it.

File: aspnet/hosting/aspnet-core-module/sample/web.config [hunk 17]
```diff
@@ -4,19 +4,19 @@
...
             resourceType="Unspecified" />
     </handlers>
-    <httpPlatform processPath="..\approot\web.cmd"
-            arguments=""
+    <aspNetCore processPath="dotnet"
+            arguments=".\MyApp.dll"
             stdoutLogEnabled="true"
-            stdoutLogFile="..\logs\stdout"
+            stdoutLogFile=".\logs\stdout"
             startupTimeLimit="3600">
```
Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module/sample/web.config]: `startupTimeLimit` has been removed from (all?) samples.

File: aspnet/tutorials/nano-server.rst [hunk 18]
```diff
@@ -52,7 +52,8 @@ If you have successfully connected then your prompt will look like this ``[10.83
 Installing the HttpPlatformHandler Module
 -----------------------------------------
 
-The :ref:`HttpPlatformHandler <http-platformhandler>` is an IIS 7.5+ module which is responsible for process management of HTTP listeners and to proxy requests to processes that it manages. At the moment, the process to install the HttpPlatformHandler Module for IIS is manual. You will need to install the latest 64-bit version of the `HttpPlatformHandler <http://www.iis.net/downloads/microsoft/HttpPlatformHandler>`_ on a regular (not Nano) machine. After installing you will need to copy the following files:
+The HttpPlatformHandler is an IIS 7.5+ module which is responsible for process management of HTTP listeners and to proxy requests to processes that it manages. At the moment, the process to install the HttpPlatformHandler Module for IIS is manual. You will need to install the latest 64-bit version of the `HttpPlatformHandler <http://www.iis.net/downloads/microsoft/HttpPlatformHandler>`_ on a regular (not Nano) machine. After installing you will need to copy the following files:
```
Comment (GuardRex) [inline aspnet/tutorials/nano-server.rst]: Nit: Comma

> After installing, you will ...

**Comment by GuardRex** [comment]
I came to the party a little late, but I hope any of those little bits are useful. :smile:

Happy Memorial Day! :rocket: :guitar: :airplane: :tada: :balloon:

File: aspnet/hosting/aspnet-core-module.rst (hunk 13 above)
Comment (rachelappel) [inline aspnet/hosting/aspnet-core-module.rst]: "Once the app_offline.htm file is removed, the next request loads the application, which then responds to requests."

File: aspnet/hosting/aspnet-core-module.rst (hunk 10 above)
Comment (rachelappel) [inline aspnet/hosting/aspnet-core-module.rst]: As per the MS style guide, we only capitalize the proper nouns in headings, so "Log creation and redirection" is the correct way.

Comment (GuardRex) [inline aspnet/hosting/aspnet-core-module.rst]: my bad ... I should have checked

File: aspnet/hosting/aspnet-core-module.rst (hunk 8 above)
Comment (shirhatti) [inline aspnet/hosting/aspnet-core-module.rst]: `shutdownTimeLimit` was an accidental omission. I will update

`processesPerApplication` was intentionally left out. It is an artifact of forking form `httpPlatformHandler`. At the moment, we don't recommend you use it with ASP.NET Core applications

File: aspnet/hosting/aspnet-core-module.rst (hunk 12 above)
Comment (shirhatti) [inline aspnet/hosting/aspnet-core-module.rst]: All valid points. :+1: Let me take another stab at this

File: aspnet/hosting/aspnet-core-module.rst (hunk 14 above)
Comment (Rick-Anderson) [inline aspnet/hosting/aspnet-core-module.rst]: They like Core Module and Tag Helper and I Can't Changer Their Mind on That.

File: aspnet/hosting/aspnet-core-module.rst (hunk 3 above)
Comment (Rick-Anderson) [inline aspnet/hosting/aspnet-core-module.rst]: I like empty string better.

File: aspnet/hosting/aspnet-core-module.rst (hunk 10 above)
Comment (Rick-Anderson) [inline aspnet/hosting/aspnet-core-module.rst]: In our unpublished style guide :--))

File: aspnet/hosting/aspnet-core-module/sample/web.config (hunk 17 above)
Comment (shirhatti) [inline aspnet/hosting/aspnet-core-module/sample/web.config]: Noted. I'll take it out of the example

**Comment by shirhatti** [comment]
@Rick-Anderson @danroth27 I made a couple more changes to configuration reference as well. I'm happy with the technical accuracy of the doc now.

I assume @Rick-Anderson will handle merging it in?

File: aspnet/hosting/aspnet-core-module.rst (hunk 8 above)
Comment (Tratcher) [inline aspnet/hosting/aspnet-core-module.rst]: Is there a specific reason not to use it?

**Comment by Rick-Anderson** [comment]
>I'm happy with the technical accuracy of the doc now.

@shirhatti  is that a squirrel?

**Comment by Rick-Anderson** [comment]
@shirhatti  I'll need help merging
D:\github\8\docs [rick-and-coremod-config ≡]> git rebase -i master
error: could not apply 6ac0715... hosting:core module config

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply 6ac07158488f36b5fa776294a4745fc6af5b829e... hosting:core module config
